src/rules/unnecessary_function_arguments.erl

```diff
@@ -3,24 +3,111 @@
 %%      ignored in all function clauses.</p>
 %%      <p>To avoid this warning, remove the unused argument(s).</p>
 %%      <p><b>Note:</b> This rule will not emit a warning if the function
-%%      implements a behaviour callback or a NIF call.</p>
+%%      implements a NIF call.</p>
 -module(unnecessary_function_arguments).
 
 -behaviour(hank_rule).
 
 -export([analyze/2, ignored/2]).
 
+%% To allow erl_syntax:syntaxTree/0 type spec
+-elvis([{elvis_style, atom_naming_convention, #{regex => "^([a-zA-Z][a-z0-9]*_?)*$"}},
+        {elvis_style, invalid_dynamic_call, disable},
+        {elvis_style, nesting_level, [6]}]).
+
+-type imp_callbacks() :: #{File :: string() => [tuple()] | ignore}.
+
 %% @private
 -spec analyze(hank_rule:asts(), hank_context:t()) -> [hank_rule:result()].
 analyze(FilesAndASTs, _Context) ->
+    ImpCallbacks = callback_usage(FilesAndASTs),
     [Result
      || {File, AST} <- FilesAndASTs,
-        not hank_utils:implements_behaviour(AST),
         not hank_utils:is_old_test_suite(File),
+        not is_unrecognized_behaviour(File, ImpCallbacks),
         Node <- AST,
         erl_syntax:type(Node) == function,
+        is_not_a_callback(ImpCallbacks, File, Node),
         Result <- analyze_function(File, Node)].
 
+%% @doc Constructs a map with the callbacks of all the files.
+%% 1. collect all the behaviors that the file implements.
+%% 2. for each one of them evaluate with BehaviourMod:behaviour_info(callbacks)
+%%    and keep the tuples in a single list.
+%%    2.1. if that call fails, look for the module/s in all the files.
+%%    2.2. if the module is not there, ignore the whole file
+%%         (we can't be sure which functions are callbacks).
+%%    2.3. If the module is there collect all callback attributes and add them
+%%         to the list obtained in the item 2.
+-spec callback_usage(hank_rule:asts()) -> imp_callbacks().
+callback_usage(FilesAndASTs) ->
+    lists:foldl(fun({File, AST}, Result) ->
+                   FoldFun =
+                       fun(Node, FileCallbacks) ->
+                          case hank_utils:node_has_attrs(Node, [behaviour, behavior]) of
+                              true ->
+                                  {_, _, _, {attribute, _, [{_, _, _, BehaviourMod}]}} = Node,
+                                  Callbacks =
+                                      try
+                                          %% For pre-compiled behaviours like OTP ones
+                                          BehaviourMod:behaviour_info(callbacks)
+                                      catch
+                                          %% Local behaviours will raise an exception
+                                          _:_ ->
```

Comment:
Can we catch the specific exception instead of any exception in general? I think it's `error:undef`